src/main/java/frc/robot/subsystems/Elevator.java

```diff
@@ -0,0 +1,156 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2018-2019 FIRST. All Rights Reserved.                        */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+package frc.robot.subsystems;
+import com.ctre.phoenix.motorcontrol.ControlMode;
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonFX;
+
+import edu.wpi.first.math.MathUtil;
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.wpilibj.DigitalInput;
+import edu.wpi.first.wpilibj.DutyCycleEncoder;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.constants.Constants;
+import frc.robot.constants.ElevatorConstants;
+import frc.robot.util.MotorFactory;
+import lib.ctre_shims.TalonEncoder;
+
+public class Elevator extends SubsystemBase {
+
+  private final WPI_TalonFX m_motor;
+  private final DigitalInput m_bottomLimitSwitch;
+  private final DigitalInput m_topLimitSwitch;
+  private final PIDController m_elevatorPID;
+  private final DutyCycleEncoder m_absoluteSpoolEncoder;     
+  private final TalonEncoder m_elevatorMotorEncoder; 
+  private double m_absEncoderZeroPositionTicks;
+  private double clampLow = -ElevatorConstants.kMotorLimit; 
+  private double clampHigh = ElevatorConstants.kMotorLimit;
+  private boolean m_enabled; 
+
+  public Elevator() {
+    m_motor = MotorFactory.createTalonFX(ElevatorConstants.kMotorPort, Constants.kRioCAN);
+    //m_motor.configForwardLimitSwitchSource(LimitSwitchSource.FeedbackConnector, LimitSwitchNormal.NormallyOpen, 0);
+    //m_motor.configReverseLimitSwitchSource(LimitSwitchSource.FeedbackConnector, LimitSwitchNormal.NormallyOpen, 0);
+    m_bottomLimitSwitch = new DigitalInput(ElevatorConstants.kBottomLimitSwitchPort); 
+    m_topLimitSwitch = new DigitalInput(ElevatorConstants.kTopLimitSwitchPort);
+    
+    m_absoluteSpoolEncoder = new DutyCycleEncoder(ElevatorConstants.kAbsEncoderPort); 
+    
+    m_elevatorMotorEncoder = new TalonEncoder(m_motor); 
+    m_elevatorMotorEncoder.setDistancePerPulse(ElevatorConstants.kDistPerMotorEncoderTick);
```

Comment:
Be careful with how you use both encoders and understand why each is necessary. The absolute encoder always tells you what angle it is facing, even if the robot is reset. But just the 0-360 angle. So if several rotations occur, it can loop on itself so it may not be enough to tell you exactly where the elevator is. The quadrature encoder in the talon does not loop, but it also does not persist on reset. So you can use the absolute encoder to persist the position, or maybe if there are limit switches that isn't necessary.

Reply:
Okay. Can I keep both of them for now?

Reply:
I think you can run it without the absolute encoder and reset the motor encoder with the limit switches,

Reply:
sounds good.